source/common/http/http2/codec_impl.cc

```diff
@@ -349,6 +354,28 @@ void ConnectionImpl::StreamImpl::encodeMetadata(const MetadataMapVector& metadat
   }
 }
 
+void ConnectionImpl::StreamImpl::processBufferedData() {
+  ENVOY_CONN_LOG(debug, "Stream {} processing buffered data.", parent_.connection_, stream_id_);
+
+  if (stream_manager_.body_buffered_ && continueProcessingBufferedData()) {
+    decodeData();
+  }
+
+  if (stream_manager_.trailers_buffered_ && continueProcessingBufferedData()) {
+    decodeTrailers();
+  }
+
+  // Activate the buffered onStreamClose if we've drained all the buffered data,
+  // or we've gotten a reset.
+  if (stream_manager_.buffered_on_stream_close_ &&
+      (!stream_manager_.hasBufferedData() || reset_reason_.has_value())) {
+    ENVOY_CONN_LOG(debug, "invoking  onStreamClose for stream: {} via processBufferedData",
+                   parent_.connection_, stream_id_);
+    // We only buffer the onStreamClose if we had no errors.
+    parent_.onStreamClose(this, 0);
```

Comment:
There is a call to parent_.onStreamClose when reset_reason_ is set in ConnectionImpl::StreamImpl::resetStream.  Can the check for reset_reason_.has_value() result in this getting called twice?  I think the answer is no, because buffered_on_stream_close_ is cleared by parent_.onStreamClose, but also implies that the reset_reason_.has_value() check is not needed here.

Reply:
Good reading, sir. yes as you said buffered_on_stream_close_ prevents us from closing the stream twice. The check of `reset_reason_.has_value()` here in not needed.